chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -118,6 +118,70 @@ impl<I: Interner> ToProgramClauses<I> for AssociatedTyValue<I> {
     }
 }
 
+impl<I: Interner> ToProgramClauses<I> for ImplTraitDatum<I> {
+    /// Given `opaque type T<..>: A + B = HiddenTy;`, we generate:
+    ///
+    /// ```notrust
+    /// AliasEq(T<..> = HiddenTy) :- Reveal.
+    /// AliasEq(T<..> = !T).
+    /// Implemented(!T: A).
+    /// Implemented(!T: B).
+    /// Implemented(!T: Send) :- Implemented(HiddenTy: Send). // For all auto traits
+    /// ```
+    /// where `!T` is the placeholder for the unnormalized type `T<..>`.
+    fn to_program_clauses(&self, builder: &mut ClauseBuilder<'_, I>) {
+        let interner = builder.db.interner();
+        let alias = AliasTy::ImplTrait(ImplTraitTy {
+            impl_trait_id: self.impl_trait_id,
+            substitution: builder.substitution_in_scope(),
+        });
+
+        let alias_ty = Ty::new(interner, alias.clone());
+
+        builder.push_binders(&self.ty, |builder, impl_trait_datum| {
+            // AliasEq(T<..> = HiddenTy) :- Reveal.
+            builder.push_clause(
+                DomainGoal::Holds(
+                    AliasEq {
+                        alias: alias.clone(),
+                        ty: impl_trait_datum.clone(),
+                    }
+                    .cast(interner),
+                ),
+                iter::once(DomainGoal::Reveal(())),
+            );
+
+            // AliasEq(T<..> = !T).
+            builder.push_fact(DomainGoal::Holds(
+                AliasEq {
+                    alias: alias.clone(),
+                    ty: alias_ty.clone(),
+                }
+                .cast(interner),
+            ));
+        });
+
+        for bound in &self.bounds {
+            // Implemented(!T: Bound).
+            builder.push_fact(bound.value.clone().into_well_formed_goal(interner));
+        }
+
+        for auto_trait_id in builder.db.auto_traits() {
```

Comment:
If we made the "hidden type" lazy, then we'd want to hold off on generating these rules unless we are asked to solve `Implemented(!T: AutoTrait)` where `AutoTrait` is, well, an auto-trait, and `!T` is some kind of impl trait placeholder type. But that would be partly done by modiying the `to_program_clauses` callback.

Reply:
Created #397 for that